packages/fiori-elements-writer/src/index.ts

```diff
@@ -0,0 +1,93 @@
+import { join } from 'path';
+import { Editor } from 'mem-fs-editor';
+import { render } from 'ejs';
+import { generate as generateUi5Project, Package } from '@sap-ux/ui5-application-writer';
+import { generate as addOdataService, OdataVersion } from '@sap-ux/odata-service-writer';
+import { getPackageJsonTasks } from './packageConfig';
+import cloneDeep from 'lodash/cloneDeep';
+import { FioriElementsApp } from './types';
+import { validateApp } from './validate';
+import { setAppDefaults, setDefaultTemplateSettings } from './data/defaults';
+import { changesPreviewToVersion, escapeFLPText, TemplateOptions } from './data/templateAttributes';
+import { extendManifestJson } from './data/manifestSettings';
+import semVer from 'semver';
+
+/**
+ * Generate a UI5 application based on the specified Fiori Elements floorplan template.
+ *
+ * @param basePath - the absolute target path where the application will be generated
+ * @param data - configuration to generate the Fiori elements application
+ * @param fs - an optional reference to a mem-fs editor
+ * @returns Reference to a mem-fs-editor
+ */
+async function generate<T>(basePath: string, data: FioriElementsApp<T>, fs?: Editor): Promise<Editor> {
+    // Clone rather than modifying callers refs
+    const feApp: FioriElementsApp<T> = cloneDeep(data);
+
+    setAppDefaults(feApp);
+
+    fs = await generateUi5Project(basePath, feApp, fs);
+
+    feApp.template.settings = setDefaultTemplateSettings(feApp.template);
+
+    // This is done after `generateUi5Project` since defaults are set if values are not provided
+    validateApp(feApp);
+
+    // Add service to the project if provided
+    if (feApp.service) {
+        if (!feApp.service.localAnnotationsName) {
+            feApp.service.localAnnotationsName = 'annotation';
+        }
```

Comment:
Does it make sense for this to be done in `setAppDefaults`?

Reply:
Yes it does, will move it thx.

Reply:
Hmm this would require some refactoring, but it makes sense if the `service` is defined.